fdbserver/worker.actor.cpp

```diff
@@ -592,6 +592,151 @@ ACTOR Future<Void> registrationClient(Reference<AsyncVar<Optional<ClusterControl
+		return true;
+	}
+
+	for (const auto& resolver : dbi.resolvers) {
+		if (resolver.address() == address) {
+			return true;
+		}
+	}
+
+	auto localityIsInPrimaryDc = [&dbInfo](const LocalityData& locality) {
+		return locality.dcId() == dbInfo->get().master.locality.dcId();
+	};
+
+	for (const auto& logSet : dbi.logSystemConfig.tLogs) {
+		for (const auto& tlog : logSet.tLogs) {
+			if (!localityIsInPrimaryDc(tlog.interf().filteredLocality)) {
+				continue;
+			}
+
+			if (tlog.interf().addresses().contain(address)) {
+				return true;
+			}
+		}
+	}
+
+	return false;
+}
+
+bool addressInDbAndPrimaryDc(const NetworkAddressList& addresses, Reference<AsyncVar<ServerDBInfo>> dbInfo) {
+	return addressInDbAndPrimaryDc(addresses.address, dbInfo) ||
+	       (addresses.secondaryAddress.present() && addressInDbAndPrimaryDc(addresses.secondaryAddress.get(), dbInfo));
+}
+
+namespace {
+
+TEST_CASE("/fdbserver/worker/addressInDbAndPrimaryDc") {
+	// Setup a ServerDBInfo for test.
+	ServerDBInfo testDbInfo;
+	LocalityData testLocal;
+	testLocal.set(LiteralStringRef("dcid"), StringRef(std::to_string(1)));
+	testDbInfo.master.locality = testLocal;
+
+	// Manually set up a master address.
+	NetworkAddress testAddress(IPAddress(0x13131313), 1);
+	testDbInfo.master.changeCoordinators =
+	    RequestStream<struct ChangeCoordinatorsRequest>(Endpoint({ testAddress }, UID(1, 2)));
+
+	// First, create a remote TLog. Although the remote TLog also uses the local address, it shouldn't be considered as
+	// in primary DC given the remote locality.
+	LocalityData fakeRemote;
+	fakeRemote.set(LiteralStringRef("dcid"), StringRef(std::to_string(2)));
+	TLogInterface remoteTlog(fakeRemote);
+	remoteTlog.initEndpoints();
+	testDbInfo.logSystemConfig.tLogs.push_back(TLogSet());
+	testDbInfo.logSystemConfig.tLogs.back().tLogs.push_back(OptionalInterface(remoteTlog));
+	ASSERT(!addressInDbAndPrimaryDc(g_network->getLocalAddress(), makeReference<AsyncVar<ServerDBInfo>>(testDbInfo)));
+
+	// Next, create a local TLog. Now, the local address should be considered as in local DC.
+	TLogInterface localTlog(testLocal);
+	localTlog.initEndpoints();
+	testDbInfo.logSystemConfig.tLogs.back().tLogs.push_back(OptionalInterface(localTlog));
+	ASSERT(addressInDbAndPrimaryDc(g_network->getLocalAddress(), makeReference<AsyncVar<ServerDBInfo>>(testDbInfo)));
+
+	// Last, use the master's address to test, which should be considered as in local DC.
+	testDbInfo.logSystemConfig.tLogs.clear();
+	ASSERT(addressInDbAndPrimaryDc(testAddress, makeReference<AsyncVar<ServerDBInfo>>(testDbInfo)));
+
+	return Void();
+}
+
+} // namespace
+
+// The actor that actively monitors the health of local and peer servers, and reports anomaly to the cluster controller.
+ACTOR Future<Void> healthMonitor(Reference<AsyncVar<Optional<ClusterControllerFullInterface>>> ccInterface,
+                                 WorkerInterface interf,
+                                 LocalityData locality,
```

Comment:
These should be references or pointers?

Reply:
Flow automatically convert such argument into a `const&`. (so explicit const& is actually not allowed).

Here is the generated code:
[[nodiscard]] Future<Void> healthMonitor( Reference<AsyncVar<Optional<ClusterControllerFullInterface>>> const& ccInterface, WorkerInterface const& interf, LocalityData const& locality, Reference<AsyncVar<ServerDBInfo>> const& dbInfo ) {
                                                                                                                        #line 681 "/root/src/foundationdb/fdbserver/worker.actor.cpp"
        return Future<Void>(new HealthMonitorActor(ccInterface, interf, locality, dbInfo));
                                                                                                                        #line 3989 "/root/build_output/fdbserver/worker.actor.g.cpp"
}